github/lib/configure_repos.rb

```diff
@@ -10,18 +10,17 @@ class ConfigureRepos
 
 
   def configure!
-    repos.each do |repo|
+    FetchRepos.new.repos.each do |repo|
       ConfigureRepo.new(repo, client, repo_overrides[repo[:full_name]]).configure!
     end
   end
 
   def list_repos
-    puts repos.to_yaml
+    puts FetchRepos.new.repos.to_yaml
   end
 
   def remove_old_webhooks!
-
-    repos.map do |repo|
+    FetchRepos.new.repos.map do |repo|
```

Comment:
There's a potential performance issue here. If we call more than one of `configure!`, `list_repos` and `remove_old_webhooks!`, we create a brand new `FetchRepos` object each time, which creates a new `Octokit::Client` each time, which means that there's probably no caching happening (so every time we make a call to `repos`, it makes a new call to GitHub's API, and as we've seen already, the calls are quite slow).

We should have a private method, like the `client` method, that sets the `FetchRepos` instance just once:

```ruby
def fetch_repos
  @fetch_repos ||= FetchRepos.new.repos
end
```

(The `||=` returns the `@fetch_repos` value if it's been set, or otherwise returns the `FetchRepos...` bit and sets the `@fetch_repos` variable to that value, i.e. it only ever calls `FetchRepos...` once. See [blog post](http://www.rubyinside.com/what-rubys-double-pipe-or-equals-really-does-5488.html)).

But we can go further. Currently, your `FetchRepos` has an `initialize` method like:

```ruby
  def initialize(client = nil)
    Octokit.auto_paginate = true
    @client = client || Octokit::Client.new(access_token: ENV.fetch("GITHUB_TOKEN"))
  end
```

As you've pointed out, this is duplication and kind of messy (if the ENV variable gets renamed or we need to pass an extra parameter to `Octokit::Client.new`, for example, then we'd currently have to update these in multiple places).

We should be able to simplify that to just:

```ruby
  def initialize(client = nil)
    @client = client
  end
```

...by changing our `fetch_repos` method to use the existing `client` method that initialises the client:

```ruby
def fetch_repos
  @fetch_repos ||= FetchRepos.new(client).repos
end
```

Reply:
I was able to implement this successfully, perfect!